source/windows/thread.c

```diff
@@ -0,0 +1,93 @@
+/*
+* Copyright 2010 - 2018 Amazon.com, Inc. or its affiliates.All Rights Reserved.
+*
+* Licensed under the Apache License, Version 2.0 (the "License").
+* You may not use this file except in compliance with the License.
+* A copy of the License is located at
+*
+*  http://aws.amazon.com/apache2.0
+*
+* or in the "license" file accompanying this file.This file is distributed
+* on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+* express or implied.See the License for the specific language governing
+* permissions and limitations under the License.
+*/
+
+#include <aws/common/thread.h>
+
+struct thread_wrapper {
+    struct aws_allocator *allocator;
+    void(*func)(void *arg);
+    void *arg;
+};
+
+static DWORD thread_wrapper_fn(LPVOID arg) {
+    struct thread_wrapper *thread_wrapper = (struct thread_wrapper *)arg;
+    thread_wrapper->func(thread_wrapper->arg);
+    aws_mem_release(thread_wrapper->allocator, thread_wrapper);
+    return 0;
+}
+
+int aws_thread_init(struct aws_thread *thread, struct aws_allocator *allocator) {
+    thread->thread_handle = 0;    
+    thread->thread_id = 0;
+    thread->allocator = allocator;
+
+    return AWS_OP_SUCCESS;
+}
+
+int aws_thread_create(struct aws_thread *thread, void(*func)(void *arg), void *context, struct aws_thread_options *options) {
+
+    SIZE_T stack_size = 0;
+
+    if (options && options->stack_size > 0) {
```

Comment:
Where do we check the detached state in the options?

Also - what are the semantics guaranteed by the detached state? Are we still required to clean up the `aws_thread`?

Reply:
I don't think the actual thread state should be in the options, but I can add the state to aws_thread for checking.  You are always required to cleanup aws_thread. The semantics should match std::thread

Reply:
See thread.h:L28. What semantics does the `aws_thread_detach_state` enumeration have, if we can ignore it?

Reply:
I don't think we ignore it. That goes in the options for creation, and will be applied then. As far as the current state of the thread, I propose, we put an instance of aws_thread_detach_state there as well.

Reply:
Where exactly do we use `aws_thread_detach_state` in the win32 thread creation path? And again, what are the semantics of setting detached state on creation?

Reply:
well, that's just an oversight on my part, not a statement about the API. 

I'll add a if(detach) CloseHandle() to the windows version.

Reply:
Sure, but what's the value of specifying detachable? I still need to clean up the `aws_thread`. What's the point in having a flag that says "Make the thing that is returned less useful, but don't make it any easier for me to use it"?

Reply:
I'm questioning the value of detach-on-create when this doesn't fully absolve the caller of any cleanup activities.

The point of the posix detached state is that you don't need to detach the thread later. On windows, this isn't an option, and instead you close the thread handle explicitly. In our API, if we are going to offer a 'detach-on-creation' option, this should - as with posix - absolve the caller from any need to perform any cleanup activities after the thread is created.

If we're not prepared to absolve the caller of any cleanup activities, then being able to pass a detach-on-creation option is useless - the _only_ function it has is to break `join`. Why would anyone pass an option that breaks a feature and does nothing else?

Reply:
sure, I get your point now. 

Maybe we can just remove it from thread creation, that's more inline with 

std::thread semantics anyways.